// marko/dist/runtime/vdom/AsyncVDOMBuilder.js
var EventEmitter = require("events-light");
var vdom = require("./vdom");
var VElement = vdom.bn_;
var VDocumentFragment = vdom.bo_;
var VComment = vdom.bp_;
var VText = vdom.bq_;
var VComponent = vdom.br_;
var VFragment = vdom.bs_;
var virtualizeHTML = vdom.bt_;
var RenderResult = require("../RenderResult");
var defaultDocument = vdom.bu_;
var morphdom = require("./morphdom");
var attrsHelper = require("./helpers/attrs");

var EVENT_UPDATE = "update";
var EVENT_FINISH = "finish";

function State(tree) {
  this.bv_ = new EventEmitter();
  this.bw_ = tree;
  this.bx_ = false;
}

function AsyncVDOMBuilder(globalData, parentNode, parentOut) {
  if (!parentNode) {
    parentNode = new VDocumentFragment();
  }

  var state;

  if (parentOut) {
    state = parentOut.L_;
  } else {
    state = new State(parentNode);
  }

  this.by_ = 1;
  this.bz_ = 0;
  this.bA_ = null;
  this.bB_ = parentOut;

  this.data = {};
  this.L_ = state;
  this.aP_ = parentNode;
  this.global = globalData || {};
  this.bC_ = [parentNode];
  this.bD_ = false;
  this.bE_ = undefined;
  this.h_ = null;

  this.m_ = null;
  this.o_ = null;
  this.aN_ = null;
}

var proto = AsyncVDOMBuilder.prototype = {
  bf_: true,
  Z_: defaultDocument,

  bc: function (component, key, ownerComponent) {
    var vComponent = new VComponent(component, key, ownerComponent);
    return this.bF_(vComponent, 0, true);
  },

  aQ_: function (component, key, ownerComponent) {
    var vComponent = new VComponent(component, key, ownerComponent, true);
    this.bF_(vComponent, 0);
  },

  bF_: function (child, childCount, pushToStack) {
    this.aP_.bG_(child);
    if (pushToStack === true) {
      this.bC_.push(child);
      this.aP_ = child;
    }
    return childCount === 0 ? this : child;
  },

  element: function (tagName, attrs, key, component, childCount, flags, props) {
    var element = new VElement(tagName, attrs, key, component, childCount, flags, props);
    return this.bF_(element, childCount);
  },

  bd_: function (tagName, attrs, key, componentDef, props) {
    return this.element(tagName, attrsHelper(attrs), key, componentDef.n_, 0, 0, props);
  },

  n: function (node, component) {
    // NOTE: We do a shallow clone since we assume the node is being reused
    //       and a node can only have one parent node.
    var clone = node.__();
    this.node(clone);
    clone.aS_ = component;

    return this;
  },

  node: function (node) {
    this.aP_.bG_(node);
    return this;
  },

  text: function (text, ownerComponent) {
    var type = typeof text;

    if (type != "string") {
      if (text == null) {
        return;
      } else if (type === "object") {
        if (text.toHTML) {
          return this.h(text.toHTML(), ownerComponent);
        }
      }

      text = text.toString();
    }

    this.aP_.bG_(new VText(text, ownerComponent));
    return this;
  },

  comment: function (comment, ownerComponent) {
    return this.node(new VComment(comment, ownerComponent));
  },

  html: function (html, ownerComponent) {
    if (html != null) {
      var vdomNode = virtualizeHTML(html, this.Z_ || document, ownerComponent);
      this.node(vdomNode);
    }

    return this;
  },

  beginElement: function (tagName, attrs, key, component, childCount, flags, props) {
    var element = new VElement(tagName, attrs, key, component, childCount, flags, props);
    this.bF_(element, childCount, true);
    return this;
  },

  bb_: function (tagName, attrs, key, componentDef, props) {
    return this.beginElement(tagName, attrsHelper(attrs), key, componentDef.n_, 0, 0, props);
  },

  bf: function (key, component, preserve) {
    var fragment = new VFragment(key, component, preserve);
    this.bF_(fragment, null, true);
    return this;
  },

  ef: function () {
    this.endElement();
  },

  endElement: function () {
    var stack = this.bC_;
    stack.pop();
    this.aP_ = stack[stack.length - 1];
  },

  end: function () {
    this.aP_ = undefined;

    var remaining = --this.by_;
    var parentOut = this.bB_;

    if (remaining === 0) {
      if (parentOut) {
        parentOut.bH_();
      } else {
        this.bI_();
      }
    } else if (remaining - this.bz_ === 0) {
      this.bJ_();
    }

    return this;
  },

  bH_: function () {
    var remaining = --this.by_;

    if (remaining === 0) {
      var parentOut = this.bB_;
      if (parentOut) {
        parentOut.bH_();
      } else {
        this.bI_();
      }
    } else if (remaining - this.bz_ === 0) {
      this.bJ_();
    }
  },

  bI_: function () {
    var state = this.L_;
    state.bx_ = true;
    state.bv_.emit(EVENT_FINISH, this.bg_());
  },

  bJ_: function () {
    var lastArray = this._last;

    var i = 0;

    function next() {
      if (i === lastArray.length) {
        return;
      }
      var lastCallback = lastArray[i++];
      lastCallback(next);

      if (!lastCallback.length) {
        next();
      }
    }

    next();
  },

  error: function (e) {
    try {
      this.emit("error", e);
    } finally {
      // If there is no listener for the error event then it will
      // throw a new Error here. In order to ensure that the async fragment
      // is still properly ended we need to put the end() in a `finally`
      // block
      this.end();
    }

    return this;
  },

  beginAsync: function (options) {
    if (this.bD_) {
      throw Error("Tried to render async while in sync mode. Note: Client side await is not currently supported in re-renders (Issue: #942).");
    }

    var state = this.L_;

    if (options) {
      if (options.last) {
        this.bz_++;
      }
    }

    this.by_++;

    var documentFragment = this.aP_.bK_();
    var asyncOut = new AsyncVDOMBuilder(this.global, documentFragment, this);

    state.bv_.emit("beginAsync", {
      out: asyncOut,
      parentOut: this
    });

    return asyncOut;
  },

  createOut: function () {
    return new AsyncVDOMBuilder(this.global);
  },

  flush: function () {
    var events = this.L_.bv_;

    if (events.listenerCount(EVENT_UPDATE)) {
      events.emit(EVENT_UPDATE, new RenderResult(this));
    }
  },

  D_: function () {
    return this.L_.bw_;
  },

  bg_: function () {
    return this.bL_ || (this.bL_ = new RenderResult(this));
  },

  on: function (event, callback) {
    var state = this.L_;

    if (event === EVENT_FINISH && state.bx_) {
      callback(this.bg_());
    } else if (event === "last") {
      this.onLast(callback);
    } else {
      state.bv_.on(event, callback);
    }

    return this;
  },

  once: function (event, callback) {
    var state = this.L_;

    if (event === EVENT_FINISH && state.bx_) {
      callback(this.bg_());
    } else if (event === "last") {
      this.onLast(callback);
    } else {
      state.bv_.once(event, callback);
    }

    return this;
  },

  emit: function (type, arg) {
    var events = this.L_.bv_;
    switch (arguments.length) {
      case 1:
        events.emit(type);
        break;
      case 2:
        events.emit(type, arg);
        break;
      default:
        events.emit.apply(events, arguments);
        break;
    }
    return this;
  },

  removeListener: function () {
    var events = this.L_.bv_;
    events.removeListener.apply(events, arguments);
    return this;
  },

  sync: function () {
    this.bD_ = true;
  },

  isSync: function () {
    return this.bD_;
  },

  onLast: function (callback) {
    var lastArray = this._last;

    if (lastArray === undefined) {
      this._last = [callback];
    } else {
      lastArray.push(callback);
    }

    return this;
  },

  C_: function (doc) {
    var node = this.bE_;
    if (!node) {
      var vdomTree = this.D_();
      // Create the root document fragment node
      doc = doc || this.Z_ || document;
      this.bE_ = node = vdomTree.bM_(doc, null);
      morphdom(node, vdomTree, doc, this.h_);
    }
    return node;
  },

  toString: function (doc) {
    var docFragment = this.C_(doc);
    var html = "";

    var child = docFragment.firstChild;
    while (child) {
      var nextSibling = child.nextSibling;
      if (child.nodeType != 1) {
        var container = docFragment.ownerDocument.createElement("div");
        container.appendChild(child.cloneNode());
        html += container.innerHTML;
      } else {
        html += child.outerHTML;
      }

      child = nextSibling;
    }

    return html;
  },

  then: function (fn, fnErr) {
    var out = this;
    var promise = new Promise(function (resolve, reject) {
      out.on("error", reject).on(EVENT_FINISH, function (result) {
        resolve(result);
      });
    });

    return Promise.resolve(promise).then(fn, fnErr);
  },

  catch: function (fnErr) {
    return this.then(undefined, fnErr);
  },

  isVDOM: true,

  c: function (componentDef, key, customEvents) {
    this.m_ = componentDef;
    this.o_ = key;
    this.aN_ = customEvents;
  }
};

proto.e = proto.element;
proto.be = proto.beginElement;
proto.ee = proto.bc_ = proto.endElement;
proto.t = proto.text;
proto.h = proto.w = proto.write = proto.html;

module.exports = AsyncVDOMBuilder;
