// marko/dist/runtime/components/legacy/defineWidget-legacy-browser.js
"use strict";
/* jshint newcap:false */

var BaseState;
var BaseComponent;
var inherit;
var jQuery = require("../jquery");
var ready = require("../ready");

var complain;
var stateToJSONDef = {
  enumerable: false,
  value: function returnSelf() {
    return this;
  }
};
function noop() {}

module.exports = function defineWidget(def, renderer) {
  def = def.Widget || def;

  if (def.___) {
    return def;
  }

  var ComponentClass = function () {};
  var proto;
  var legacyInit;

  if (typeof def === "function") {
    proto = def.prototype;
    legacyInit = def;
  } else if (typeof def === "object") {
    proto = def;
    legacyInit = def.init;
  } else {
    throw TypeError();
  }

  ComponentClass.prototype = proto;

  // We don't use the constructor provided by the user
  // since we don't invoke their constructor until
  // we have had a chance to do our own initialization.
  // Instead, we store their constructor in the "initComponent"
  // property and that method gets called later inside
  // init-components-browser.js
  function Component(id, doc) {
    BaseComponent.call(this, id, doc);
  }

  if (!proto.___) {
    // Inherit from Component if they didn't already
    ComponentClass.prototype = Object.create(BaseComponent.prototype);
    for (var propName in proto) {
      if (proto.hasOwnProperty(propName)) {
        ComponentClass.prototype[propName] = proto[propName];
      }
    }
  }

  // The same prototype will be used by our constructor after
  // we he have set up the prototype chain using the inherit function
  proto = Component.prototype = ComponentClass.prototype;
  proto.ao_ = true;

  proto.constructor = def.constructor = Component;

  Object.defineProperty(proto, "state", {
    get: function () {
      var raw = this.L_ && this.L_._v_;

      if (raw && !raw.toJSON) {
        Object.defineProperty(this.L_._v_, "toJSON", stateToJSONDef);
      }

      return raw;
    },
    set: function (newState) {
      newState = newState || {};
      // eslint-disable-next-line no-constant-condition


      this.setState(newState);
    }
  });

  Object.defineProperty(proto, "__document", {
    get: function () {
      return this.Z_;
      // eslint-disable-next-line no-constant-condition
    }
  });

  Object.defineProperty(proto, "el", {
    get: function () {

      return this.az_;
      // eslint-disable-next-line no-constant-condition
    }
  });

  // get legacy methods

  proto.aA_ = proto.onRender;
  Object.defineProperty(proto, "onRender", {
    get: noop,
    set: function (v) {
      proto.aA_ = v;
    }
  });

  proto.aB_ = proto.onUpdate;
  Object.defineProperty(proto, "onUpdate", {
    get: function () {
      return modernMountOrUpdate;
    },
    set: function (v) {
      proto.aB_ = v;
    }
  });

  proto.aC_ = proto.onDestroy;
  Object.defineProperty(proto, "onDestroy", {
    get: function () {
      return modernOnDestory;
    },
    set: function (v) {
      proto.aC_ = v;
    }
  });

  proto.getWidget = proto.getComponent;
  proto.getWidgets = proto.getComponents;
  proto.onMount = modernMountOrUpdate;

  if (legacyInit) {
    proto.aD_ = legacyInit;
  }

  // convert legacy to modern
  proto.aE_ = proto.update;
  proto.update = function () {
    if (this.V_) {

      this.destroy = modernOnDestory;
      // eslint-disable-next-line no-constant-condition

      this.V_ = false;
    }
    this.aF_ = true;
    if (this.az_) {
      this.onBeforeUpdate && this.onBeforeUpdate();
    }

    this.aE_();
    this.aF_ = false;
  };

  function modernMountOrUpdate() {
    var self = this;
    var el = this.p_["@_wbind"];
    var prevEl = this.az_;

    if (prevEl !== el) {
      this.az_ = el;

      if (prevEl) {
        this.onBeforeDestroy && this.onBeforeDestroy();
        this.aC_ && this.aC_();
        this.aG_ = undefined;
        this.removeAllListeners();
      }

      if (el) {
        this.aD_ && this.aD_(this.widgetConfig || {});
        this.aA_ && this.aA_({ firstRender: true });
        this.aG_ = legacyRender;

        // eslint-disable-next-line no-constant-condition
        el.__widget = this;
      }
    } else if (el) {
      if (prevEl) {
        this.aB_ && this.aB_();
      }

      if (this.aH_) {
        this.aA_ && this.aA_({ firstRender: false });
      }
    }

    this.ay_ = this.S_;
    this.S_ = null;
    this.aH_ = false;
  }

  function legacyRender() {
    if (!this.aF_) {
      this.onBeforeUpdate && this.onBeforeUpdate();
    }

    this.aH_ = true;
  }

  function modernOnDestory() {
    if (this.az_) {
      this.onBeforeDestroy && this.onBeforeDestroy();
      this.aC_ && this.aC_();
      this.az_ = null;
    }
  }

  // Set a flag on the constructor function to make it clear this is
  // a component so that we can short-circuit this work later
  Component.___ = true;

  function State() {
    BaseState.apply(this, arguments);
  }
  inherit(State, BaseState);
  proto._e_ = State;

  jQuery.patchComponent(window.$, proto, true /* don't throw error until used if `$` is missing*/
  );

  ready.patchComponent(proto);

  if (!renderer) {
    renderer = ComponentClass.renderer || ComponentClass.prototype.renderer;
    if (renderer) {
      // Legacy support
      var createOut = renderer.createOut;
      if (typeof renderer !== "function") {
        var rendererObject = renderer;
        renderer = function (input, out) {
          var rendererFunc = rendererObject.renderer || rendererObject.render;
          rendererFunc(input, out);
        };
        renderer.createOut = createOut;
      }

      renderer.render = function (input) {
        var out = createOut();
        renderer(input, out);
        return out.end();
      };
    }
  }

  if (renderer) {
    // Add the rendering related methods as statics on the
    // new component constructor function
    Component.renderer = proto._p_ = renderer;
    Component.render = renderer.render;
    Component.renderSync = renderer.renderSync;
  }

  // eslint-disable-next-line no-constant-condition
  Component._isWidget = true;


  return Component;
};

BaseState = require("../State");
BaseComponent = require("../Component");
inherit = require("raptor-util/inherit");
