// marko/dist/runtime/components/ComponentsContext.js
"use strict";

var GlobalComponentsContext = require("./GlobalComponentsContext");

function ComponentsContext(out, parentComponentsContext) {
  var globalComponentsContext;
  var componentDef;

  if (parentComponentsContext) {
    globalComponentsContext = parentComponentsContext.k_;
    componentDef = parentComponentsContext._R_;

    var nestedContextsForParent;
    if (!(nestedContextsForParent = parentComponentsContext._S_)) {
      nestedContextsForParent = parentComponentsContext._S_ = [];
    }

    nestedContextsForParent.push(this);
  } else {
    globalComponentsContext = out.global.h_;
    if (globalComponentsContext === undefined) {
      out.global.h_ = globalComponentsContext = new GlobalComponentsContext(out);
    }
  }

  this.k_ = globalComponentsContext;
  this.h_ = [];
  this.A_ = out;
  this._R_ = componentDef;
  this._S_ = undefined;
  this.r_ = parentComponentsContext && parentComponentsContext.r_;
}

ComponentsContext.prototype = {
  B_: function (doc) {
    var componentDefs = this.h_;

    ComponentsContext._T_(componentDefs, doc);

    this.A_.emit("_U_");

    // Reset things stored in global since global is retained for
    // future renders
    this.A_.global.h_ = undefined;

    return componentDefs;
  }
};

function getComponentsContext(out) {
  return out.h_ || (out.h_ = new ComponentsContext(out));
}

module.exports = exports = ComponentsContext;

exports.q_ = getComponentsContext;
